src/modbus.c

```diff
@@ -1559,6 +1728,140 @@ int modbus_report_slave_id(modbus_t *ctx, int max_dest, uint8_t *dest)
+        file_no, req);
+
+    req[req_length++] = read_addr >> 8;
+    req[req_length++] = read_addr & 0x00ff;
+    req[req_length++] = read_nb >> 8;
+    req[req_length++] = read_nb & 0x00ff;
+
+    rc = send_msg(ctx, req, req_length);
+    if (rc > 0) {
+        int offset;
+
+        rc = _modbus_receive_msg(ctx, rsp, MSG_CONFIRMATION);
+        if (rc == -1)
+            return -1;
+
+        rc = check_confirmation(ctx, req, rsp, rc);
+        if (rc == -1)
+            return -1;
+
+        rc /= 2; /* rc is in byte, we count in uint16-steps */
+        offset = ctx->backend->header_length;
+        if (rc > (read_nb + 1)) { /* One register is extra for size and ref-type */
+            if (ctx->debug) {
+                fprintf(stderr, "ERROR Too many data for destination (%d > %d)\n", rc,
+                        read_nb);
+            }
+            rc = read_nb;
+        }
+
+        /* If rc is negative, the loop is jumped ! */
+        for (i = 0; i < rc; i++) {
+            /* shift reg hi_byte to temp OR with lo_byte */
+            dest[i] = (rsp[offset + 2 + (i << 1)] << 8) | rsp[offset + 3 + (i << 1)];
+        }
+    }
+
+    return rc;
+}
+
+/* Write General Reference ( aka as Write File ) writes nb registers (16-bit)
+ * from an offset ( also 16-bit stepping ) of a given filenumber*/
+/* This implements only the simple case with one subrequest. More complex can be
+ * created with raw-message. */
+int modbus_write_general_reference(modbus_t *ctx, int file_no, int write_addr,
+                                   int write_nb, const uint16_t *src)
+
+{
+    int rc;
+    int req_length;
+    int i;
+    int byte_count;
+    uint8_t req[MAX_MESSAGE_LENGTH];
+    uint8_t rsp[MAX_MESSAGE_LENGTH];
+
+    if (write_nb > MODBUS_MAX_READ_REGISTERS) {
+        if (ctx->debug) {
+            fprintf(stderr, "ERROR Too many registers requested (%d > %d)\n",
+                    write_nb, MODBUS_MAX_READ_REGISTERS);
+        }
+        errno = EMBMDATA;
+        return -1;
+    }
+    byte_count = SUB_REQUEST_LENGHT + (write_nb * 2);
+    req_length = ctx->backend->build_request_basis(
+        ctx, MODBUS_FC_WRITE_GENERAL_REFERENCE, ((byte_count << 8) | 0x06),
+        file_no, req);
+
+    req[req_length++] = write_addr >> 8;
+    req[req_length++] = write_addr & 0x00ff;
+    req[req_length++] = write_nb >> 8;
+    req[req_length++] = write_nb & 0x00ff;
+
+    for (i = 0; i < write_nb; i++) {
+        req[req_length++] = src[i] >> 8;
+        req[req_length++] = src[i] & 0xff;
+    }
+
+    rc = send_msg(ctx, req, req_length);
+    if (rc > 0) {
+
+        rc = _modbus_receive_msg(ctx, rsp, MSG_CONFIRMATION);
+        if (rc == -1)
+            return -1;
+
+        rc = check_confirmation(ctx, req, rsp, rc);
+
+        if (rc == -1)
+            return -1;
+
+        rc /= 2; /* rc is in byte, we count in uint16-steps */
+
+#if 0 // Does it make sense to copy back the response to the caller ? It's a
+      // copy of the request
```

Comment:
I wondered whether the standard enforces to simply mirror the request content back to the caller or if it would be possible to e.g. return the previous data of the file at this position. But after I've read the standard again, I think it's intension really it to reflect the request back. So I think we can just remove this here.